### DISCUSSION TITLE
Preparing v6

### DISCUSSION THREAD

**Discussion by drbyte** [spatie/laravel-permission#2317] [General]
Some of the pending PR's involve breaking or nearly-breaking changes, so I'm considering tagging v6 for them.
Should some of these be kept in v5 anyway?

- #2280 - fix withCount support -- affects Role model. 
- #2230 - hint model properties -- affects Role and Permission models
- #2266 - add types to function signatures -- affects a lot of files (needs merge conflicts fixed, and patterns applied to all files added by other PRs)

- #2252 - custom wildcard verification support - not seriously breaking, but extended classes should be redone.

Probably not breaking, but touch main Registrar file:
- #2309 - lazily bind dependencies (performance improvement)
- #2316 - extract getPermissionsWithRoles method

Teams -- @erikn69 should both of these be merged as-is? I know some time has passed without much comment on them, so does that mean they're less critical because they're easily addressed in userland instead of here?
- #1935 - teams trait and team deletion logic
- #2221 - detaching user models on teams

**Comment by erikn69** [comment]
Hi, I also think that Php8 syntax should be adopted for v6, so we leave support for 7.3, 7.4

**Teams**

- #2221 
It's a bug fix, I added a test to demonstrate the problem, it can be merged as-is on V5(#2220)
- #1935 it is opcional trait, if you want to keep the package as light as possible, you can ignore it, because it can be handled manually in each implementation. Add it in case someone has that problem, the PR can be tagged(Maybe a discussion should be added)

**Others**

- #2252 It makes no difference to me in the version that is added or if it is not added, it was only to help #2110, 
Also it **avoids** calling `if(config('permission.enable_wildcard_permission', false))` on every `hasPermissionTo`
- #2309 Looks great to me, it could fixes https://github.com/spatie/laravel-permission/discussions/2310#discussioncomment-4757929, but I don't know if it affects any scenario outside of my knowledge
- #2316 It is not a breaking change for V5, but looks like helps only on specific cases, 
- #2230 It is only hints, it doesn't change functionality, it could be added on V5
- #2266 V6

**Reply by drbyte** [reply]
Your comments about v5 are noted. I'll probably merge these today or tomorrow:
- #2221
- #2252
- #2309
- #2316
- #2230
- perhaps #1935 after additional review

**Reply by erikn69** [reply]
looks good to me 👍

- #1935 i think useless on most of the cases, that's why I put it optional

**Reply by drbyte** [reply]
@erikn69 thanks for the feedback! Greatly appreciated!

**Reply by erikn69** [reply]
> #2309 Looks great to me, it could fixes https://github.com/spatie/laravel-permission/discussions/2310#discussioncomment-4757929, but I don't know if it affects any scenario outside of my knowledge

It was a breaking change #2318, #2319, i did make a commit for fix that(https://github.com/erikn69/laravel-permission/commit/d0f799c69a713f3045a1d87d655960e24f1cd4dc) but i don't know if is better revert the PR

**Comment by parallels999** [comment]
- #2089 UUIDs/ULIDs/GUIDs support on V6 - change data type on findById is a breaking change for somebody overwriting that function, if that is true, maybe for V6 release
- [create_permission_tables.php](https://github.com/spatie/laravel-permission/blob/main/database/migrations/create_permission_tables.php.stub) use anonymous classes in migrations on V6

**Reply by erikn69** [reply]
If necessary UUID support I can do a rebase
For anonymous classes Laravel 7.x, 8.x should be dropped

**Comment by drbyte** [comment]
I'm reluctant to completely drop all Laravel 7 and 8, and PHP7 support if some of the open PRs should be applied for sites still running there. 
Obviously, supporting legacy dependencies has a limit. 
So, that begs the question, what open PRs should be merged into v5 before we tag v6? (ie: avoiding the need to backport)
EDIT: I'm working through Eric's comments now.

**Reply by parallels999** [reply]
>what open PRs should be merged into v5 before we tag v6?

From the list i think 
- #2230
- #2221
- #2316
- #2309

**Reply by drbyte** [reply]
@parallels999 thanks for the feedback! Greatly appreciated!

**Comment by marcoraddatz** [comment]
In the last few months there have been two changes in minor updates that needed to be reverted. Although I love this package and I am grateful to everyone who contributes to it, I would suggest to be more careful regarding new features.

So I would suggest better to release more major versions whenever there is even the slightest chance of breaking something, or even start with pre-release versions, that can be tested within test environments. Since this package is not in line with Laravel versioning, is there anything against releasing 1+ major versions per year?

I'd put following PRs in v6:
- https://github.com/spatie/laravel-permission/pull/2309 (the follow up)
- https://github.com/spatie/laravel-permission/pull/2266
- https://github.com/spatie/laravel-permission/pull/2291 (should require tests?!)

**Reply by parallels999** [reply]
- #2291 
i think these must be closed and look for another approach

**Comment by drbyte** [comment]
#2280 and #2321 are now merged into `6.x` / `dev-main` for testing.

**Reply by parallels999** [reply]
@drbyte what about #2089
>  https://github.com/spatie/laravel-permission/discussions/2317#discussioncomment-4883429

**Comment by parallels999** [comment]
@drbyte what about [spatie/laravel-package-tools](https://github.com/spatie/laravel-package-tools) for the service provider?

**Reply by drbyte** [reply]
Perhaps. I wonder if it just creates more overhead, potentially slowing things down?

What specific benefit do you see it providing?

**Reply by parallels999** [reply]
yes you could be right, never mind

**Reply by drbyte** [reply]
I did explore whether the "install" command might be beneficial, but it's not really much different than simply publishing assets. Most of the rest of the "install" process actually requires programmer intervention. So, I'm still leaning away from it, to reduce complexity.

**Comment by systemsolutionweb** [comment]
@drbyte hi, any release date for this? any RC?

**Reply by drbyte** [reply]
At this point it's essentially stable. Just not officially released.
You can use it now as version `^6.0-dev`.
Do consult the `upgrading.md` file for things related to migrations and tests.

**Comment by drbyte** [comment]
I'm looking for arguments for-and-against **defaulting to UUID** in `v6`.

- FOR: simpler for those who need it, because it's already done.
- FOR: a lot of systems use UUID by default
- AGAINST: it's overkill if you don't have a lot of permissions/roles
- AGAINST: takes up more database space and RAM and cache
- AGAINST: it's a lot harder to use `tinker` and specify a role/permission `id` in bespoke commands
- AGAINST: converting old data to uuid is a bit of a nuisance, but probably a minor issue

I'm on the fence.

**Comment by drbyte** [comment]
@erikn69 do you mind reviewing the latest updates to the `uuid.md` documentation file, to be sure nothing's been missed in the edits I've merged and tweaked?

**Reply by erikn69** [reply]
Everything looks in order to me

**Comment by drbyte** [comment]
@erikn69 what are your thoughts on defaulting to using uuid on the pivot table morph key column, in model_has_roles and model_has_permissions?
Like, if we defaulted to uuid, then it would support uuid/ulid/int out-of-the-box with no changes required.
One potential "downside" is that uuid sets the field as varchar instead of as bigUnsignedInt ... but does that actually cause any performance degradation by the lookup when matching on foreign key models? 
In the simple test I've modeled, MariaDB 10.6 isn't complaining about mismatched column types in the foreign relationship, and lookups function fine, but it's a very small dataset I'm testing with.
Curious your thoughts.

**Reply by erikn69** [reply]
I have always maintained the idea of ​​having the data types as they should be, I have them as big int, and I would not change them, I think it should adapt to the needs of the majority, if it does not affect performance I do not see problems

#2513 is incomplete and the complexity is increasing.

**Reply by drbyte** [reply]
> https://github.com/spatie/laravel-permission/pull/2513 is incomplete and the complexity is increasing.

agreed.

**Reply by drbyte** [reply]
> I have always maintained the idea of ​​having the data types as they should be, I have them as big int, and I would not change them

Ya, me too.

> I think it should adapt to the needs of the majority, if it does not affect performance I do not see problems

I see the value in uuid, but I think int will always be more performant.  


I prefer ulid over uuid in my apps, but uuid takes up more chars so any package should specify uuid, which will then accommodate both uuid and ulid.

**Comment by drbyte** [comment]
Update on v6 release: 
I had hoped to battle-test v6 on a complex new project I've been helping build ... but that project got half-done and the stakeholders decided to put the brakes on for several months, for various reasons, some of which are sensible, others are not. But I digress.

So, I hope to release v6 very soon. Within a couple weeks.
